src/storage-client/src/collections_controller.rs

```diff
@@ -0,0 +1,1164 @@
+
+            if PartialOrder::less_equal(&collection.implied_capability, &new_read_capability) {
+                let mut update = ChangeBatch::new();
+                update.extend(new_read_capability.iter().map(|time| (time.clone(), 1)));
+                std::mem::swap(&mut collection.implied_capability, &mut new_read_capability);
+                update.extend(new_read_capability.iter().map(|time| (time.clone(), -1)));
+                if !update.is_empty() {
+                    read_capability_changes.insert(id, update);
+                }
+            }
+
+            collection.read_policy = policy;
+        }
+
+        drop(collections);
+
+        if !read_capability_changes.is_empty() {
+            self.update_read_capabilities(&mut read_capability_changes);
+        }
+    }
+
+    fn acquire_read_holds(
+        &mut self,
+        desired_holds: Vec<(GlobalId, Antichain<Self::Timestamp>)>,
+    ) -> Vec<(GlobalId, Antichain<Self::Timestamp>)> {
+        let collections = self.collections.lock().expect("lock poisoned");
+
+        // Clone for logging!
+        let cloned_desired_holds = desired_holds.clone();
+
+        // We advance the holds by our current since frontier. Can't
+        // acquire holds for times that have been compacted away!
+        let advanced_holds = desired_holds.into_iter().map(|(id, mut desired_hold)| {
+            let collection = collections.get(&id).expect("missing collection");
+            desired_hold.join_assign(&collection.implied_capability);
+            (id, desired_hold)
+        });
+
+        // Report back the holds that we could acquire.
+        let res = advanced_holds.clone().collect();
+
+        let mut updates = advanced_holds
+            .into_iter()
+            .map(|(id, hold)| {
+                let mut changes = ChangeBatch::new();
+                changes.extend(hold.into_iter().map(|time| (time, 1)));
+
+                (id, changes)
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        drop(collections);
+
+        self.update_read_capabilities(&mut updates);
+
+        info!(desired_holds = ?cloned_desired_holds, acquired_holds = ?res, "acquire_read_holds");
+
+        res
+    }
+
+    fn release_read_holds(&mut self, holds: Vec<(GlobalId, Antichain<Self::Timestamp>)>) {
+        info!(holds = ?holds, "release_read_holds");
+
+        let mut updates = holds
+            .into_iter()
+            .map(|(id, hold)| {
+                let mut changes = ChangeBatch::new();
+                changes.extend(hold.into_iter().map(|time| (time, -1)));
+
+                (id, changes)
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        self.update_read_capabilities(&mut updates)
+    }
+
+    fn downgrade_read_holds(
+        &mut self,
+        changes: Vec<(
+            GlobalId,
+            Antichain<Self::Timestamp>,
+            Antichain<Self::Timestamp>,
+        )>,
```

Comment:
These differential APIs can introduce annoying bugs when different parts of the system disagree about what the previous state was. If the client sends us an `old` value that for whatever reason doesn't match whatever is in the `read_capabilities` `MutableAntichain`, we will retract the wrong thing and the read frontier will get stuck at the time we didn't retract instead. Or alternatively, we notice the invalid state of the `MutableAntichain` and panic instead, which is not great either.

What do you think of remembering for each read hold the held time instead and producing the retraction that way? That would require introducing some identifier for read holds (a `ReadHoldId`) but we might want to do this anyway because we could then also keep track of additional metadata for debugging (like who acquired a read hold for what purpose).

(Fine to not consider this in the scope of this PR though!)

Reply:
Yeah, I have this under the umbrella of "more type-safe API", what you wrote about elsewhere that we plumb around a token that represents the hold, and it's opaque to the user. Very handwave-y right now!